### PR TITLE
Added iCub V3 support to iKin.

### PR DESCRIPTION
**Pull request by ale-git**

### CODE REVIEW

File: src/libraries/iKin/src/iKinFwd.cpp [hunk 1]
```diff
@@ -1566,18 +1566,45 @@ iCubTorso::iCubTorso(const string &_type) : iKinLimb(string(""))
...
+    {
+        version = 1.0;
+    }
+
     Matrix H0(4,4);
     H0.zero();
+    H0(3, 3) = 1.0;
+
+    if (version < 3.0)
+    {
```
Comment (pattacini) [inline src/libraries/iKin/src/iKinFwd.cpp]: Indentation problem.

File: src/libraries/iKin/src/iKinFwd.cpp [hunk 2]
```diff
@@ -1566,18 +1566,45 @@ iCubTorso::iCubTorso(const string &_type) : iKinLimb(string(""))
...
 }
+    else // version >= 3.0
+    {
+        H0(0, 2) = 1.0;
+        H0(1, 1) = -1.0;
+        H0(2, 0) = 1.0;
+        setH0(H0);
+
+        const double mm2m = 0.001;
+        pushLink(new iKinLink(72.5*mm2m, 0.0*mm2m, -90.0*CTRL_DEG2RAD, 0.0*CTRL_DEG2RAD));
```
Comment (pattacini) [inline src/libraries/iKin/src/iKinFwd.cpp]: Please revise in order not to use `mm2m` for being consistent with the rest of the file.

File: src/libraries/iKin/src/iKinFwd.cpp [hunk 3]
```diff
@@ -1566,18 +1566,45 @@ iCubTorso::iCubTorso(const string &_type) : iKinLimb(string(""))
 /************************************************************************/
 void iCubTorso::allocate(const string &_type)
 {
+    double version;
+    size_t underscore = getType().find('_');
```
Comment (pattacini) [inline src/libraries/iKin/src/iKinFwd.cpp]: Please revise to be consistent with the formatting of the file.
For example, no spaces around symbols `=`, `,` ...

File: src/libraries/iKin/src/iKinFwd.cpp [hunk 4]
```diff
@@ -1566,18 +1566,45 @@ iCubTorso::iCubTorso(const string &_type) : iKinLimb(string(""))
...
+    {
+        H0(0, 2) = 1.0;
+        H0(1, 1) = -1.0;
+        H0(2, 0) = 1.0;
+        setH0(H0);
+
+        const double mm2m = 0.001;
+        pushLink(new iKinLink(72.5*mm2m, 0.0*mm2m, -90.0*CTRL_DEG2RAD, 0.0*CTRL_DEG2RAD));
+        pushLink(new iKinLink(0.0*mm2m, 0.0*mm2m, 90.0*CTRL_DEG2RAD, -90.0*CTRL_DEG2RAD));
+        pushLink(new iKinLink(-4.86151*mm2m, -263.77*mm2m, -90.0*CTRL_DEG2RAD, 180.0*CTRL_DEG2RAD));
```
Comment (pattacini) [inline src/libraries/iKin/src/iKinFwd.cpp]: Please align function parameters as in the rest of the file.

File: src/libraries/iKin/src/iKinFwd.cpp [hunk 5]
```diff
@@ -2128,6 +2200,46 @@ void iCubLeg::allocate(const string &_type)
         pushLink(new iKinLink(       0.0,      0.0,  M_PI/2.0,       0.0,  -42.0*CTRL_DEG2RAD,  21.0*CTRL_DEG2RAD));
         pushLink(new iKinLink(  -0.06805,   0.0035,      M_PI,       0.0,  -24.0*CTRL_DEG2RAD,  24.0*CTRL_DEG2RAD));
     }
+    else if (getType() == "right_v3")
+    {
+        H0.zero();
+        H0(0, 0) = 1.0;
+        H0(1, 2) = 1.0;
+        H0(2, 1) = -1.0;
+        H0(0, 3) = -21.6*0.001;
```
Comment (pattacini) [inline src/libraries/iKin/src/iKinFwd.cpp]: numbers directly in SI.

File: src/libraries/iKin/src/iKinFwd.cpp [hunk 6]
```diff
@@ -2227,6 +2339,33 @@ void iCubEye::allocate(const string &_type)
         pushLink(new iKinLink(    0.0,   0.034, -M_PI/2.0,       0.0, -35.0*CTRL_DEG2RAD, 15.0*CTRL_DEG2RAD));
         pushLink(new iKinLink(    0.0,     0.0,  M_PI/2.0, -M_PI/2.0, -50.0*CTRL_DEG2RAD, 50.0*CTRL_DEG2RAD));
     }
+    else if (getType() == "left_v3" || getType() == "right_v3")
```
Comment (pattacini) [inline src/libraries/iKin/src/iKinFwd.cpp]: Use parentheses around logical operands as in the rest of the files.

File: src/libraries/iKin/src/iKinFwd.cpp [hunk 7]
```diff
@@ -1566,18 +1566,45 @@ iCubTorso::iCubTorso(const string &_type) : iKinLimb(string(""))
 /************************************************************************/
 void iCubTorso::allocate(const string &_type)
 {
+    double version;
+    size_t underscore = getType().find('_');
+    if (underscore != string::npos)
+    {
```
Comment (pattacini) [inline src/libraries/iKin/src/iKinFwd.cpp]: The convention in this file is not to use curly brackets `{` for single operations.

**Comment by traversaro** [comment]
@ale-git It would be good to document somewhere from which version of the iCub V3 Creo assembly these DH parameters have been extracted. These DH have been extracted manually or through the URDF and the urdf2dh iDynTree utility?

**Comment by ale-git** [comment]
@traversaro: we don't have yet the URDF model, thus the DH parameters have been manually extracted from the CAD model. About the model version, Alberto Parmiggiani tells me that we can specify the SVN index of the CAD model in the comments.

@pattacini: this branch was based on the master by mistake, please reject this pull request. I'm going to make the pull request from the right branch. I've fixed the style too.

**Comment by pattacini** [comment]
@ale-git I had already rebased the PR on `devel`, which is quite easy to do from within the PR UI: see [this][1].

Closing anyway.

[1]: https://blog.github.com/2016-08-15-change-the-base-branch-of-a-pull-request

**Pull request closed without merging**